### ISSUE TITLE
`cast_possible_truncation` could check whether floating point values are manually truncated (round, floor, ceil...) prior to cast

### ISSUE THREAD

**Issue by ijijn** [rust-lang/rust-clippy#13257]
### Summary

`cast_possible_truncation` can be overly zealous about casting floating point values to integers when said values have just been explicitly "rounded" in the code, so that their values would fall comfortably within the range of the target type. However, if the target type is still incapable of holding all possible values then this lint should apply regardless.

Ideally, clippy would check to see whether some kind of manual truncation/rounding, etc. has occurred to prevent these false positives.

The example given uses `f64` but this logic would naturally extend to other float/integer input/output pairs too in the presence of manually-implemented truncation.

:)

### Lint Name

cast_possible_truncation

### Reproducer

I tried this code:

```rust
#![warn(clippy::pedantic)]

fn main() {
    let x = 16578.8f64;
    // rounding f64 effectively produces an "i53", so these casts should always succeed perfectly
    let y = x.round() as i64;
    let z = x.ceil() as i64;
    let w = x.floor() as i64;
    dbg!(x, y, z, w);
}

```

I saw this happen:

```
casting `f64` to `i64` may truncate the value
if this is intentional allow the lint with `#[allow(clippy::cast_possible_truncation)]` ...
for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cast_possible_truncation
`#[warn(clippy::cast_possible_truncation)]` implied by `#[warn(clippy::pedantic)]`
```
for each of the three casts.

I expected to see this happen:

_no warnings emitted_

### Version

```text
rustc 1.82.0-nightly (730d5d409 2024-08-10)
binary: rustc
commit-hash: 730d5d4095a264ef5f7c0a0781eea68c15431d45
commit-date: 2024-08-10
host: x86_64-pc-windows-msvc
release: 1.82.0-nightly
LLVM version: 19.1.0
```

**Comment by ijijn** [comment]
Hmm... it's not quite so simple is it, since `f64::MAX.round()` is much larger than `i64::MAX` because of the whole exponent business, but in most sane use cases it seems unlikely to come up. In any event, as with other issues surround `cast_possible_truncation`, detecting any constants that fall within certain ranges would be very helpful in reducing false positives.
